go/workflow.go

```diff
@@ -3,45 +3,42 @@ package app
 import (
 	"time"
 
+	"go.temporal.io/sdk/temporal"
 	"go.temporal.io/sdk/workflow"
-	"go.uber.org/multierr"
 )
 
-func BreakfastWorkflow(ctx workflow.Context) (err error) {
+func BreakfastWorkflow(ctx workflow.Context, parallelCompensations bool) (err error) {
 	options := workflow.ActivityOptions{
 		StartToCloseTimeout: time.Second * 5,
+		RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 1},
 	}
 
 	ctx = workflow.WithActivityOptions(ctx, options)
 
-	err = workflow.ExecuteActivity(ctx, GetBowl).Get(ctx, nil)
-	if err != nil {
-		return err
-	}
+	var compensations Compensations
 
 	defer func() {
+		// Defer is at the top so that it is executed regardless of which step might fail.
 		if err != nil {
-			errCompensation := workflow.ExecuteActivity(ctx, PutBowlAway).Get(ctx, nil)
-			err = multierr.Append(err, errCompensation)
+			// activity failed, and workflow context is canceled
+			disconnectedCtx, _ := workflow.NewDisconnectedContext(ctx)
+			compensations.Compensate(disconnectedCtx, parallelCompensations)
 		}
 	}()
 
-	err = workflow.ExecuteActivity(ctx, AddCereal).Get(ctx, nil)
+	err = workflow.ExecuteActivity(ctx, GetBowl).Get(ctx, nil)
+	compensations.AddCompensation(PutBowlAway)
```

Comment:
Based on samples from other languages, I don't think you want to add compensation for an activity until _after_ you know it succeeds.

Reply:
sorry, I'm confused, isn't that what's happening here?